daemon/api_snaps.go

```diff
@@ -260,6 +261,44 @@ func (inst *snapInstruction) holdLevel() snapstate.HoldLevel {
 	}
 }
 
+func cleanSnapshotOptions(snapshotOpts map[string]*snap.SnapshotOptions) {
+	for name, options := range snapshotOpts {
+		if options.Unset() {
+			delete(snapshotOpts, name)
+		}
+	}
+}
+
+// cleanAndValidateSnapshotOptions cleans and validates the snapshot options.
+//
+// With default marshalling, some permutations of valid JSON definitions of snapshot-options e.g.
+//   - `"snapshot-options": { "snap1": {} }`
+//   - `"snapshot-options": { "snap1": {exclude: []} }`
+//
+// results in a pointer to SnapshotOptions object with a nil or zero length exclusion list
+// which in turn will be marshalled to JSON as `options: {}` when we rather want it omitted.
+// The cleaning step ensures that we only populate map entries for snapshot options that contain
+// usable content that we want to be marshalled downstream.
+func (inst *snapInstruction) cleanAndValidateSnapshotOptions() error {
+	if inst.SnapshotOptions == nil {
+		return nil
+	}
+	if inst.Action != "snapshot" {
+		return fmt.Errorf("snapshot-options can only be specified for snapshot action")
+	}
+	cleanSnapshotOptions(inst.SnapshotOptions)
```

Comment:
Putting the cleanSnapshotOptions here means that input that is arguably  invalid (putting in an empty snapshot options for a snap that is not part of the snapshot set) is not flagged. This is functionally fine but I wonder if instead we should error as this may indicate a bug in the API caller - i.e. they should just not populate the api fields incorrectly even if in this case it has no effect?